src/providers/completer/reference.ts

```diff
@@ -1,114 +1,109 @@
         })
+        // Compile the suggestion object to array
+        return Object.keys(this.suggestions).map(key => this.suggestions[key])
     }
 
-    getReferenceItems(content: string) {
-        const itemReg = /(?:\\label(?:\[[^[\]{}]*\])?|(?:^|[,\s])label=){([^}]*)}/gm
-        const items: {[key: string]: ReferenceEntry['item']} = {}
-        content = stripComments(content, '%')
-        const noELContent = content.split('\n').filter(para => para !== '').join('\n')
-        while (true) {
-            const result = itemReg.exec(content)
-            if (result === null) {
-                break
-            }
-            if (!(result[1] in items)) {
-                const prevContent = noELContent.substring(0, noELContent.substring(0, result.index).lastIndexOf('\n') - 1)
-                const followLength = noELContent.substring(result.index, noELContent.length).split('\n', 4).join('\n').length
-                const positionContent = content.substring(0, result.index).split('\n')
-                items[result[1]] = {
-                    reference: result[1],
-                    text: `${noELContent.substring(prevContent.lastIndexOf('\n') + 1, result.index + followLength)}\n...`,
-                    position: new vscode.Position(positionContent.length - 1, positionContent[positionContent.length - 1].length)
+    update(file: string, content: string) {
+        const refs = this.getReferenceFromAST(latexParser.parse(content), content.split('\n'))
+        this.extension.manager.cachedContent[file].element.reference = refs
+    }
+
+    private getReferenceFromAST(ast: latexParser.LatexAst | latexParser.Node, lines: string[], nextNode?: latexParser.Node): vscode.CompletionItem[] {
+        let refs: vscode.CompletionItem[] = []
+        let label = ''
+        if (ast.kind === 'command' && ast.name === 'label') {
+            // \label{some-text}
+            label = (ast.args.filter(arg => arg.kind === 'arg.group')[0].content[0] as latexParser.TextString).content
+        } else if (ast.kind === 'text.string' && ast.content === 'label=' && nextNode !== undefined) {
+            // label={some=text}
+            label = ((nextNode as latexParser.Group).content[0] as latexParser.TextString).content
+        }
+        if (label !== '' && (ast.kind === 'command' || ast.kind === 'text.string')) {
+            refs.push({
+                label,
+                kind: vscode.CompletionItemKind.Reference,
+                documentation: lines.slice(ast.location.start.line - 2, ast.location.end.line + 4).join('\n'),
+                range: new vscode.Range(ast.location.start.line - 1, ast.location.start.column,
+                                        ast.location.end.line - 1, ast.location.end.column)
+            })
+            // Here we abuse the definition of range to store the location of the reference definition
+            return refs
+        }
+        if (ast.hasOwnProperty('content') && Array.isArray(ast['content'])) {
```

Comment:
I have added `hasContent` and `hasContentArray`. Please use them. See [link](https://github.com/tamuratak/latex-utensils/blob/master/src/latex_parser_types.ts#L223).